[#11] 요청 데이터 logging 기능 구현

hyena0608 opened the pull request:
## 🌱 작업 사항
### aop 의존성 추가
- Spring AOP를 이용하기 위해 spring-boot-starter-aop 의존성을 추가하였습니다.
### 요청 데이터 consume 문제 해결
- POST 요청을 받은 HttpServletRequest의 body 한 번 이용시 body 데이터가 소비되는 문제를 해결했습니다.
  - HttpServletRequestWrapper를 상속받아 커스텀하여 body 데이터를 필드에 저장했습니다.
  - Filter를 구현하여 요청 받은 Content-Type이 `multipart/form-data`가 아닐 시에 커스텀한 RequestWrapper 객체를 이용합니다.
### AOP를 이용한 logging
- `com.prgrms.prolog.domain..api.*Controller` 에 해당하는 컨트롤러만 요청 데이터를 로깅합니다.
```
2023-01-14 21:15:08.480  INFO 13827 --- [nio-8080-exec-2] com.prgrms.prolog.global.log.LogAspect  :30 : Request -----> TestResponse com.prgrms.prolog.domain.test.api.TestController.createTest(TestRequest) -> (POST /test), ({"name":"테스트","age":10,"testArray":["테스트 배열1","테스트 배열2","테스트 배열3","테스트 배열4"]})
```
- 응답 데이터 로깅을 넣지 않았습니다. 
  - 다음과 같은 이유를 고려하였습니다.
    - 많은 응답 데이터 로깅으로 인해 비용 증가 가능성이 있습니다.
    - 성공에 대한 응답 데이터를 로깅할 이유를 찾지 못했습니다.
    - 실패에 대한 응답 데이터 로깅은 전역 예외 처리에서 역할을 갖고 있습니다.

### 요청 데이터 logging 리팩터링
- LogAspect, 커스텀 HttpServletRequestWrapper, 커스텀 Filter을 삭제한다.
- Spring에서 제공하는 AbstractRequestLoggingFilter를 상속받아 요청 데이터 로깅을 처리한다.

## 🦄 관련 이슈
close #11

hyena0608 committed:
config: Spring AOP 의존성 추가

hyena0608 committed:
add: Aspectj auto proxy 어노테이션 추가

hyena0608 committed:
feat: 요청 커스텀 wrapper 및 요청 커스텀 필터 구현

hyena0608 committed:
feat: 요청 데이터 로깅 기능 구현

JoosungKwon commented on src/main/java/com/prgrms/prolog/global/log/CustomHttpServletRequestWrapper.java:
```diff
@@ -0,0 +1,77 @@
+public class CustomHttpServletRequestWrapper extends HttpServletRequestWrapper {
+
+	private final Map<String, String[]> paramMap = new HashMap<>();
+	private final byte[] cachedBody;
```
혹시 paramMap하고 cachedBody하고 무슨 차이인가요..?

hyena0608 replied:
요청을 받았을 때 다음과 같이 데이터를 보관하게 됩니다 ! 

- paramMap
  - QueryString
  - Content-Type: application/x-www-form-urlencoded
  - ...
- cachedBody
  - request body

CustomHttpServletRequestWrapper를 구현한 이유는 유실될 수 있는 request body를 cachedBody에 보관해두기 위함입니다.

paramMap에 유실되지 않는 데이터를 보관한 이유는 getReader() 메서드에서  HttpServletRequest를 매개변수로 전달해주지 않고 보다 편리하게 처리하기 위해서였습니다. 🫣

hyena0608 replied:
요청을 받았을 때 다음과 같이 데이터를 보관하게 됩니다 ! 

- paramMap
  - QueryString
  - Content-Type: application/x-www-form-urlencoded
  - ...
- cachedBody
  - request body

CustomHttpServletRequestWrapper를 구현한 이유는 유실될 수 있는 request body를 cachedBody에 보관해두기 위함입니다.

paramMap에 유실되지 않는 데이터를 보관한 이유는 getReader() 메서드에서  HttpServletRequest를 매개변수로 전달해주지 않고 보다 편리하게 처리하기 위해서였습니다. 🫣

JoosungKwon replied:
아하 헤더와 비슷하다고 생각하면 되겠네요! 그렇다면 밑에 getCachedBody 메서드에서 paramMap의 isEmpty 메서드를 호출했는데 이유가 뭔가요? paramMap이 비어 있을 수 있나요?

hyena0608 replied:
```java
public String getCachedBody() throws JSONException {
  if (paramMap.isEmpty() && cachedBody.length != 0) {
	  return new JSONObject(getReader().lines().collect(Collectors.joining())).toString();
  }
```

httpMethod: POST
Content-Type: application/json

형태로 요청이 왔을 때 paramMap은 비어있는 상태가 맞습니다.
요청에 paramMap으로 들어올 데이터가 넘어오지 않았기 때문입니다.

---
ps.

현재 애플리케이션에서 문제가 될 부분은 없다고 판단했습니다.
하지만 주성님이 보신 것처럼 현재 코드를 보고 어떤 의미를 갖는지 알 수 없는 부분이 있네요 😨

생각해보면
생성자 내에 request를 가지고 Content-Type을 확인하거나
여쭤보신 것처럼 매개변수로 HttpServeltRequest를 받아와서 메서드 내부에서 더 명확하게 표현할 수 있겠네요. 🤔

리팩터링 해보도록 하겠습니다 !

JoosungKwon replied:
아 제가 착각했네요!! parmMap에 헤더 영역이 전부 들어간다고 생각했습니다!! httpServlet에 익숙하지 않아서 변수명이 헷갈렸네요 😅
지금도 충분히 괜찮은 것 같습니다!! 👍 
ps. 쿼리 파리미터와 리퀘스트 바디 둘다 들어오면 어떻게 되나요?

Sinyoung3016 commented:
구현하신 로깅 과정에 대해서 설명 부탁드려도 될까요?🥹

hyena0608 committed:
refactor: 요청 데이터 Logging 리팩터링

- AOP를 이용한 LogAspect를 삭제한다.
- 커스텀 HttpServletRequestWrapper를 삭제한다.
- 커스텀 Filter를 삭제한다.
- AbstractRequestLoggingFilter를 상속받아 구현한다.

hyena0608 committed:
update: 이전 버전 요청 데이터 logging 삭제

- AOP를 이용한 LogAspect를 삭제한다.
- 커스텀 HttpServletRequestWrapper를 삭제한다.
- 커스텀 Filter를 삭제한다.

hyena0608 committed:
update: AOP 설정 삭제

- gradle에 AOP 의존을 삭제한다.
- main Application에 AOP 어노테이션을 삭제한다.
